flax/experimental/nnx/nnx/compat/module.py

```diff
@@ -0,0 +1,226 @@
+  type_counter: defaultdict[type, int] = dataclasses.field(
+    default_factory=lambda: defaultdict(lambda: 0)
+  )
+
+
+@dataclasses.dataclass
+class ModuleContext(threading.local):
+  parent_stack: list[tp.Optional[CompactContext]] = dataclasses.field(
+    default_factory=lambda: [None]
+  )
+
+
+MODULE_CONTEXT = ModuleContext()
+
+
+@dataclasses.dataclass
+class Scope(Object):
+  rngs: rnglib.Rngs
+
+
+@tp.runtime_checkable
+class _HasSetup(tp.Protocol):
+  def setup(self) -> None: ...
+
+
+class ModuleMeta(nnx_module.ModuleMeta):
+  if not tp.TYPE_CHECKING:
+
+    def __call__(cls, *args, **kwargs):
+      return _module_meta_call(cls, *args, **kwargs)
+
+
+def _module_meta_call(cls: tp.Type[M], *args, **kwargs) -> M:
+  # compact behavior
+  parent_ctx = MODULE_CONTEXT.parent_stack[-1]
+  parent = None
+  module: M
+
+  if parent_ctx is not None:
+    if 'parent' in kwargs:
+      parent = kwargs.pop('parent')
+      if parent is not None:
+        raise ValueError(
+          f"'parent' can only be set to None, got {type(parent).__name__}"
+        )
+      name = None
+    else:
+      type_index = parent_ctx.type_counter[cls]
+      parent_ctx.type_counter[cls] += 1
+
+      # define the name
+      if 'name' in kwargs:
+        name = kwargs.pop('name')
+        if not isinstance(name, str):
+          raise ValueError(f"'name' must be a 'str', got {type(name).__name__}")
+      else:
+        name = f'{cls.__name__}_{type_index}'
+
+      parent = parent_ctx.module
+
+      if hasattr(parent, name):
+        module = getattr(parent, name)
+        return module
+  else:
+    name = None
+
+  module = nnx_module.ModuleMeta.__call__(cls, *args, **kwargs)
+  module.scope = None
+
+  if parent is not None:
+    assert name is not None
+    setattr(parent, name, module)
+    # adopt the parent scope
+    module.scope = parent.scope
+
+  if dataclasses.is_dataclass(module):
+    if isinstance(module, _HasSetup):
+      module.setup()
+
+  return module
+
+
+class ModuleBase:
+  if tp.TYPE_CHECKING:
+    scope: Scope | None
+
+
+@tpe.dataclass_transform(field_specifiers=(dataclasses.field,))
+class Module(nnx_module.Module, ModuleBase, metaclass=ModuleMeta):
+  def _set_scope(self, scope: Scope | None):
+    """Recursively sets the scope for the Module and its children."""
+    for _, value in graph.iter_graph(self):
+      if isinstance(value, Module):
+        value.scope = scope
+
+  @property
+  def init(self: M) -> M:
+    """Calls a method in initialization mode.
+
+    When a method is called using ``init``, the ``is_initializing`` method
+    will return ``True``. This is useful to implement Modules that support
+    lazy initialization.
+
+    Example::
+
+      >>> from flax.experimental import nnx
+      >>> from flax.experimental.nnx import compat as nnc
+      >>> import jax
+      >>> import jax.numpy as jnp
+      ...
+      >>> class Linear(nnc.Module):
+      ...   def __init__(self, dout, rngs: nnx.Rngs):
```

Comment:
For clarification:
* if we define an `__init__` method, we will be able to instantiate module parameters using the `.init` method?
* if we define a [`setup`](https://github.com/google/flax/pull/3921/files#diff-6b8d092686d8c9b95f179640fa4095d53562e065a929b93c0010d7631ab38f2dR29) method or wrap `__call__` with [`compact`](https://github.com/google/flax/pull/3921/files#diff-6b8d092686d8c9b95f179640fa4095d53562e065a929b93c0010d7631ab38f2dR47), we will be able to instantiate the module parameters by [calling the module](https://github.com/google/flax/pull/3921/files#diff-6b8d092686d8c9b95f179640fa4095d53562e065a929b93c0010d7631ab38f2dR59) on a sample input and invoking shape inference?
* the module parameters that are instantiated are bound to the module so they can be dot-accessed, which is different from Linen where they are returned separately as a variable dict?
* Instead of defining an `__init__` method, can we define a `setup` method or wrap `__call__` with `compact` to use the `.init` method as well?